ipme.c

```diff
@@ -52,20 +52,35 @@ int ipme_init()
   byte_copy(&ix.ip,4,"\0\0\0\0");
   if (!ipalloc_append(&ipme,&ix)) { return 0; }
   if ((s = socket(AF_INET,SOCK_STREAM,0)) == -1) return -1;
- 
-  len = 256;
-  for (;;) {
-    if (!stralloc_ready(&buf,len)) { close(s); return 0; }
-    buf.len = 0;
+
+  ifc.ifc_buf = 0;
+  ifc.ifc_len = 0;
+
+  /* first pass: just ask what the correct length for all addresses is */
+  len = 0;
+  if (ioctl(s,SIOCGIFCONF,&ifc) >= 0 && ifc.ifc_len > 0) { /* > is for System V */
+    if (!stralloc_ready(&buf,ifc.ifc_len)) { close(s); return 0; }
     ifc.ifc_buf = buf.s;
-    ifc.ifc_len = len;
-    if (ioctl(s,SIOCGIFCONF,&ifc) >= 0) /* > is for System V */
-      if (ifc.ifc_len + sizeof(*ifr) + 64 < len) { /* what a stupid interface */
-        buf.len = ifc.ifc_len;
-        break;
-      }
-    if (len > 200000) { close(s); return -1; }
-    len += 100 + (len >> 2);
+    if (ioctl(s,SIOCGIFCONF,&ifc) >= 0)
```

Comment:
The interface is: set the length to 0, call the ioctl, and it will tell you how much buffer you need. Allocate it, set the length properly, call ioctl again. This way you have the right buffer size after 2 calls. DJB did a loop and some testing to see if he had enough. This is still the fallback path in case this does not work as intended, but your testing showed that this seems to work pretty well (or the fallback is triggered and still works fine).